activerecord/lib/active_record/relation.rb

```diff
@@ -298,6 +298,35 @@ def many?
       limit_value ? to_a.many? : size > 1
     end
 
+    # Returns a cache key that can be used to identify the records fetched by
+    # this query. The cache key is built with a fingerprint of the sql query,
+    # the number of records matched by the query and a timestamp of the last
+    # updated record. When a new record comes to match the query, or any of
+    # the existing records is updated or deleted, the cache key changes.
+    #
+    #   Product.where("name like ?", "%Cosmic Encounter%").cache_key
+    #   => "products/query-1850ab3d302391b85b8693e941286659-1-20150714212553907087000"
+    #
+    # Under the hood this triggers two SQL queries:
+    #
+    #    SELECT COUNT(*) FROM "products" WHERE (name like '%Cosmic Encounter%')
+    #    SELECT MAX("products"."updated_at") FROM "products" WHERE (name like '%Cosmic Encounter%')
+    #
+    # You can also pass a custom timestamp column to fetch the timestamp of the
+    # last updated record.
+    #
+    #   Product.where("name like ?", "%Game%").cache_key(:last_reviewed_at)
+    def cache_key(timestamp_column = :updated_at)
+      query_signature = Digest::MD5.hexdigest(to_sql)
```

Comment:
@egilburg thanks, for your feedback, here are my thoughts:

1. Although MD5 can have some issues for cryptography I think it is fine for fingerprinting. It's somewhat less CPU intensive and it's what the Asset Pipeline uses for fingerprints.

2. That's a good idea, although I'll have to check if I can create the query in a DB independent manner.

3. It will fail hard with an missing column exception. I think that's the right behaviour.

4. The string length it's not an issue, AFAICT. Average key length ~ 75 chars (depending on the model name), while memcached keys can be up to 255 chars. I'd prefer to optimise for the readability of the key.